From 1dacdf951f524f287e3ea8130fd4bea3936f55af Mon Sep 17 00:00:00 2001
From: Richard Zhu <r65037@freescale.com>
Date: Fri, 25 Jul 2008 15:08:20 +0800
Subject: [PATCH] ENGR00083066 SD Card Random Read CRC error

The driver would report random Read CRC(-84) error for some SD
cards, use software workaround to fix this issue.
Ignore this error when this issue occured after cmd51 is issued.
Refer to the ENGR00070542 and ENGR00070108 CRs.

Signed-off-by: Richard Zhu <r65037@freescale.com>
---
 drivers/mmc/host/mxc_mmc.c |   11 +++++++++--
 1 files changed, 9 insertions(+), 2 deletions(-)

diff --git a/drivers/mmc/host/mxc_mmc.c b/drivers/mmc/host/mxc_mmc.c
index ce013c0..0959785 100644
--- a/drivers/mmc/host/mxc_mmc.c
+++ b/drivers/mmc/host/mxc_mmc.c
@@ -632,8 +632,15 @@ static int mxcmci_cmd_done(struct mxcmci_host *host, unsigned int stat)
 			while (!(__raw_readl(host->base + MMC_STATUS) &
 				 (STATUS_BUF_READ_RDY | STATUS_READ_OP_DONE))) ;
 
+			pr_debug("status is 0x%x\n",
+				 __raw_readl(host->base + MMC_STATUS));
 			/* read 32 bit data */
 			temp_data = __raw_readl(host->base + MMC_BUFFER_ACCESS);
+			if (SD_APP_SEND_SCR == cmd->opcode) {
+				pr_debug("CMD51 read out 0x%x\n", temp_data);
+				if (temp_data == 0xFFFFFFFF)
+					temp_data = 0;
+			}
 			if (no_of_bytes >= 4) {
 				*buf++ = temp_data;
 				no_of_bytes -= 4;
@@ -658,7 +665,8 @@ static int mxcmci_cmd_done(struct mxcmci_host *host, unsigned int stat)
 				     host->base + MMC_STATUS);
 		} else if (status & STATUS_READ_CRC_ERR) {
 			pr_debug("%s: Read CRC error occurred\n", DRIVER_NAME);
-			data->error = -EILSEQ;
+			if (SD_APP_SEND_SCR != cmd->opcode)
+				data->error = -EILSEQ;
 			__raw_writel(STATUS_READ_CRC_ERR,
 				     host->base + MMC_STATUS);
 		}
@@ -1001,7 +1009,6 @@ static void mxcmci_set_ios(struct mmc_host *mmc, struct mmc_ios *ios)
 			prescaler = 0;
 
 		/* clk_dev =1, CLK_DIV = ipg_perclk/2 */
-
 		while (prescaler <= 0x800) {
 			for (clk_dev = 1; clk_dev <= 0xF; clk_dev++) {
 				int x;
-- 
1.5.4.4

